File: python-patterns/app/racing_agents.py
import typing
import restate
from pydantic import BaseModel

from restate import Service, Context, RestateDurableCallFuture, RunOptions
from app.util.litellm_call import llm_call


class Question(BaseModel):
    message: str = "What's the best approach to learn machine learning?"


# <start_here>
racing_agent = Service("RacingAgent")


@racing_agent.handler()
async def run(ctx: Context, query: Question) -> str | None:
    """Run two approaches in parallel and return the fastest response."""
    # Start both service calls concurrently
    slow_response = ctx.service_call(think_longer, arg=query)
    quick_response = ctx.service_call(respond_quickly, arg=query)

    done, pending = await restate.wait_completed(slow_response, quick_response)

    # cancel the pending calls
    for f in pending:
        call_future = typing.cast(RestateDurableCallFuture, f)
        ctx.cancel_invocation(await call_future.invocation_id())

    # return the fastest result
    return await done[0]


# <end_here>


@racing_agent.handler()
async def think_longer(ctx: Context, req: Question) -> str | None:
    output = await ctx.run_typed(
        "Deep analysis",
        llm_call,
        RunOptions(max_attempts=3),
        messages=f"Analyze this thoroughly: {req}",
    )
    return output.content


@racing_agent.handler()
async def respond_quickly(ctx: Context, req: Question) -> str | None:
    output = await ctx.run_typed(
        "Quick response",
        llm_call,
        RunOptions(max_attempts=3),
        messages=f"Quick answer: {req}",
    )
    return output.content

File: typescript-patterns/src/racing-agents.ts
/**
 * Competitively racing agents
 *
 * Run two approaches in parallel and return the fastest response.
 * Cancel the slower tasks to save resources.
 */
import * as restate from "@restatedev/restate-sdk";
import { Context, RestatePromise } from "@restatedev/restate-sdk";
import llmCall from "./utils/llm";
import { zodPrompt } from "./utils/utils";

const examplePrompt = "What's the best approach to learn machine learning?";

// <start_here>
async function run(
  ctx: Context,
  { message }: { message: string },
): Promise<string> {
  // Start both service calls concurrently
  const slowCall = ctx.serviceClient(racingAgent).thinkLonger({ message });
  const slowResponse = slowCall.map((res) => ({ tag: "slow", res }));

  const fastCall = ctx.serviceClient(racingAgent).respondQuickly({ message });
  const fastResponse = fastCall.map((res) => ({ tag: "fast", res }));

  const pending = [slowResponse, fastResponse];

  // Wait for the first one to complete
  const { tag, res } = await RestatePromise.any(pending);

  if (tag === "fast") {
    console.log("Quick response won the race!");
    const slowInvocationId = await slowCall.invocationId;
    ctx.cancel(slowInvocationId);
  } else {
    console.log("Deep analysis won the race!");
    const quickInvocationId = await fastCall.invocationId;
    ctx.cancel(quickInvocationId);
  }

  return res ?? "LLM gave no response";
}
// <end_here>

async function thinkLonger(
  ctx: Context,
  { message }: { message: string },
): Promise<string> {
  const { text } = await ctx.run(
    "Deep analysis",
    async () => llmCall(`Analyze this thoroughly: ${message}`),
    { maxRetryAttempts: 3 },
  );
  return text;
}

async function respondQuickly(
  ctx: Context,
  { message }: { message: string },
): Promise<string> {
  const { text } = await ctx.run(
    "Quick response",
    async () => llmCall(`Quick answer: ${message}`),
    { maxRetryAttempts: 3 },
  );
  return text;
}

const racingAgent = restate.service({
  name: "RacingAgent",
  handlers: {
    run: restate.createServiceHandler({ input: zodPrompt(examplePrompt) }, run),
    thinkLonger,
    respondQuickly,
  },
});

export default racingAgent;
